src/features/cart/ShoppingCartTest.ts

```diff
@@ -0,0 +1,185 @@
+    if(isItem.quantity>1){
+        isItem.quantity--,
+        console.log("One Item Remove ",isItem)
+
+    }
+    else if(isItem.quantity=1){
+        cart.items=cart.items.filter((item)=>item.id!==isItem.id)
+        // cart.items.filter((item)=>item.id !==isItem.id)
+        console.log("No Item From this Cart")
+
+    }
+    else{
+        console.log("No item form this Cart",id)
+    }
+}
+export const addDiscountCode=(discountCode:string):void=>{
+    //Find the Code 
+    const checkCode=discountCodes.find((item)=>item.code==discountCode)
+    if(checkCode){
+        if(checkCode.type="cart"){
+            //all Ready Use this Code Or not 
+            const isApplied=cart.discount.includes(checkCode.code)
+            if(!isApplied){
+                cart.discount.push(checkCode.code)
+            }
+            else{
+                console.log("This Code allready Exits in the Cart",isApplied)
+                return
+            }
+        }
+        else{
+            const product=cart.items[cart.items.findIndex((item)=>item.id==checkCode.itemId)]
+            //All ready Use or not
+            const isApplid=product.discount.includes(checkCode.code)
+            if(!isApplid){
+                product.discount.push(checkCode.code)
+                console.log("Descount Code is item",product)
+
+            }else{
+                console.log("All Ready Use this code before")
+                return
+            }
+
+        }
+
+    }
+    else{
+        console.log("There Is No Code Here ",discountCode)
+    }
+
+}
+//Get Single Card Price 
+export const singlePrice=(item:Item):number=>{
+    const codeApplied = item.discount.length > 0;
+  let totalPrice = 0;
+  if (codeApplied) {
+    // get code and then its value from discountCodes
+    const codeValue = discountCodes.filter((code) =>
+      item.discount.includes(code.code)
+    );
+    console.log(codeValue);
+    const totalDiscount = codeValue.reduce(
+      (total, item) => total + item.value,
+      0
+    );
+    console.log(totalDiscount);
+    totalPrice =(item.price * item.quantity) - ((item.price * item.quantity * totalDiscount) / 100);
+  } else {
+    totalPrice = item.quantity * item.price;
+  }
+  return totalPrice;
+}
+export const grandTotal=():number=>{
+  // check if any item has discounted price 
+  const haveDiscount = cart.items.filter(item => item.discount.length > 0);
+  let totalDiscountedItemPrice : number = 0;
+  if(haveDiscount){
+    const discountedItemPrice : number[] = [];
+    // get discounted price of these items 
+    haveDiscount.forEach(item => {
+     const price =  singlePrice(item);
+     discountedItemPrice.push(price)
+    })
+    totalDiscountedItemPrice = discountedItemPrice.reduce((total, item) => total + item, 0)
+  }
+  const codeApplied = cart.discount.length > 0;
+  let totalPrice = 0;
+  if (codeApplied) {
+    const codeValue = discountCodes.filter((code) =>
+      cart.discount.includes(code.code)
+    );
```

Comment:
We could use an alternative approach same as above, and use a reusable private function to calculate total discount given an array of discount codes. 

But since this is a working code, you do not have to modify it. I just wanted to say it for future reference.